pkg/plugins/resources/cargopackage/main.go

```diff
@@ -0,0 +1,159 @@
+package cargopackage
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/mitchellh/mapstructure"
+	"github.com/sirupsen/logrus"
+	"github.com/updatecli/updatecli/pkg/plugins/utils/version"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+// CargoPackage defines a resource of type "cargopackage"
+type CargoPackage struct {
+	spec Spec
+	//options []remote.Option
+	// versionFilter holds the "valid" version.filter, that might be different from the user-specified filter (Spec.VersionFilter)
+	versionFilter version.Filter
+	foundVersion  version.Version
+	packageData   PackageData
+}
+
+type PackageVersion struct {
+	Name     string `json:"name"`
+	Version  string `json:"vers"`
+	Yanked   bool   `json:"yanked"`
+	Checksum string `json:"chsum"`
+}
+
+type PackageData struct {
+	Name     string
+	Versions []PackageVersion
+}
+
+// New returns a reference to a newly initialized CargoPackage object from a cargopackage.Spec
+// or an error if the provided Spec triggers a validation error.
+func New(spec interface{}) (*CargoPackage, error) {
+
+	newSpec := Spec{}
+
+	err := mapstructure.Decode(spec, &newSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	newFilter, err := newSpec.VersionFilter.Init()
+	if err != nil {
+		return nil, err
+	}
+
+	newResource := &CargoPackage{
+		spec:          newSpec,
+		versionFilter: newFilter,
+	}
+
+	return newResource, nil
+}
+
+// Validate tests that tag struct is correctly configured
+func (cp *CargoPackage) Validate() error {
```

Comment:
```
func (s *Spec) Validate() error {
```

IHMO this function should be moved to the file `spec.go` and validate the spec object only

It should only be called from the function `New(spec interface{}) (*CargoPackage, error) {`